src/scripts/interactions.js

```diff
@@ -0,0 +1,216 @@
+        const target = new Matrix4().copy(intersection.object.matrixWorld);
+        const transformMatrix = new Matrix4().multiplyMatrices(pointerInverse, target);
+        data = {
+          object: intersection.object,
+          transformMatrix,
+          matrixAutoUpdate: intersection.object.matrixAutoUpdate
+        };
+      }
+      intersection.object.matrixAutoUpdate = false;
+      dragAndDrop.set(inputSource, data);
+    } else if (interactions.select_start) {
+      interactions.select_start(intersection);
+    }
+  }
+});
+const handleSelectEnd = handlerCommon((intersection, inputSource) => {
+  // Handle the end of dragging
+  const data = dragAndDrop.get(inputSource);
+  if (data) {
+    const dragend = data.object[Interactions].drag_end;
+    data.object.matrixAutoUpdate = data.matrixAutoUpdate;
+    dragAndDrop.delete(inputSource);
+    if (dragend) {
+      dragend();
+    }
+  }
+
+  const interactions = intersection.object[Interactions];
+  if (interactions) {
+    // Handle the end of selection
+    if (interactions.select_end) {
+      interactions.select_end(intersection);
+    }
+  }
+});
+const handleSelect = handlerCommon((intersection) => {
+  const interactions = intersection.object[Interactions];
+  if (interactions) {
+    if (interactions.select) {
+      interactions.select(intersection);
+    }
+  }
+});
+
+export function setupInteractions() {
+  inputSources = XR.session.getInputSources();
+  console.log(inputSources);
+  XR.session.addEventListener('inputsourceschange', handleInputSourcesChange);
+
+  XR.session.addEventListener('select', handleSelect);
+  XR.session.addEventListener('selectstart', handleSelectStart);
+  XR.session.addEventListener('selectend', handleSelectEnd);
+}
+
+const raycaster = new Raycaster();
+function raycast(xrRay) {
+  const { scene } = getCurrentScene();
+
+  const trMatrix = new Matrix4().fromArray(xrRay.matrix);
+
+  // Transformed ray matrix from the current scene matrix world
+  const rMatrix = new Matrix4().multiplyMatrices(scene.matrixWorld, trMatrix);
+
+  raycaster.set(
+    new Vector3().setFromMatrixPosition(rMatrix),
+    new Vector3(0, 0, -1).transformDirection(rMatrix)
+      .normalize()
+  );
+  const intersections = raycaster.intersectObjects(scene.children, true);
+  for (const intersection of intersections) {
+    intersection.point.applyMatrix4(scene.matrixWorld);
+  }
+  return intersections;
+}
+
+const lastObjects = new Map();
+function updateInputSource(inputSource, ray) {
+  // Handle Drag and Drop
+  if (dragAndDrop.has(inputSource)) {
+    const { object, transformMatrix } = dragAndDrop.get(inputSource);
+    const newMatrix = new Matrix4().multiplyMatrices(new Matrix4().fromArray(ray.matrix), transformMatrix);
+    if (object[Interactions].drag) {
+      object[Interactions].drag(newMatrix);
+    } else {
+      object.matrix = newMatrix;
+      object.updateMatrixWorld(true);
+    }
+  }
+
```

Comment:
The logic that handles a case where I have selected (held down) an object and moved the ray away from it (ray is no longer intersecting the object but I am still holding down and selecting) is missing. The result is that when I select and hold an object then move the ray away (without dragging) it stops selecting that object entirely and raycasts other objects. I handled this before by checking to see if the user was raycasting another object but was still selecting a previous object. I would then just return the previous object in the list of interactions and ignore everything else until the user released the object.

I was able to repro this error by selecting a box in the homeroom and moving the ray away.

Expected behavior:
1. User hovers over object
    - hover_start called and isIntersected is set
    - object set as prev intersection for next frame
2. User selects object
    - select_start is called and isSelected is set
    - every frame will by default return this object as the sole intersection
3. User moves cursor away from object and ray is no longer intersecting (hovering) that object in physical space
    - object remains selected and is the only intersection returned
    - hover_exit is called and isIntersected is unset
4. User releases object
    - raycaster resumes calculating and returning intersections as normal
    - select_end is called and isSelected is unset

Reply:
I didn't make these changes to the interactions themselves, but I think I got what you want added to the TriggerMesh.  This behavior is on TriggerMesh but only on the TriggerMesh callbacks like exit, hover, select, and release.